runtime-sdk/modules/evm/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+    fn create<C: TxContext>(
+        ctx: &mut C,
+        value: U256,
+        init_code: Vec<u8>,
+        gas_price: U256,
+        gas_limit: u64,
+    ) -> Result<Vec<u8>, Error> {
+        let caller = Self::derive_caller(ctx)?;
+        Self::do_evm(caller, value, gas_price, gas_limit, ctx, |exec| {
+            let address = exec.create_address(evm::CreateScheme::Legacy {
+                caller: caller.into(),
+            });
+            (
+                exec.transact_create(caller.into(), value.into(), init_code, gas_limit),
+                address.as_bytes().to_vec(),
+            )
+        })
+    }
+
+    fn call<C: TxContext>(
+        ctx: &mut C,
+        address: H160,
+        value: U256,
+        data: Vec<u8>,
+        gas_price: U256,
+        gas_limit: u64,
+    ) -> Result<Vec<u8>, Error> {
+        let caller = Self::derive_caller(ctx)?;
+        Self::do_evm(caller, value, gas_price, gas_limit, ctx, |exec| {
+            exec.transact_call(caller.into(), address.into(), value.into(), data, gas_limit)
+        })
+    }
+
+    fn peek_storage<C: Context>(ctx: &mut C, address: H160, index: H256) -> Result<Vec<u8>, Error> {
+        let store = storage::PrefixStore::new(ctx.runtime_state(), &crate::MODULE_NAME);
+        let storages = storage::PrefixStore::new(store, &state::STORAGES);
+        let s = storage::TypedStore::new(storage::PrefixStore::new(storages, &address));
+
+        let result: H256 = s.get(&index).unwrap_or_default();
+
+        Ok(result.as_bytes().to_vec())
+    }
+
+    fn peek_code<C: Context>(ctx: &mut C, address: H160) -> Result<Vec<u8>, Error> {
+        let store = storage::PrefixStore::new(ctx.runtime_state(), &crate::MODULE_NAME);
+        let codes = storage::TypedStore::new(storage::PrefixStore::new(store, &state::CODES));
+
+        Ok(codes.get(&address).unwrap_or_default())
+    }
+}
+
+impl Module {
+    const EVM_CONFIG: Config = Config::istanbul();
+
+    fn do_evm<C, F, V>(
+        source: H160,
+        value: U256,
+        gas_price_in: U256,
+        gas_limit: u64,
+        ctx: &mut C,
+        f: F,
+    ) -> Result<V, Error>
+    where
+        F: FnOnce(
+            &mut StackExecutor<'static, MemoryStackState<'_, 'static, evm_backend::Backend<'_, C>>>,
+        ) -> (evm::ExitReason, V),
+        C: Context,
+    {
+        // Make sure that gas_price >= min_gas_price.
+        let params = Self::params(ctx.runtime_state());
+        let gas_price: primitive_types::U256 = gas_price_in.into();
+        let min_gas_price: primitive_types::U256 = params.min_gas_price.into();
+        if gas_price < min_gas_price {
+            return Err(Error::GasPriceTooLow);
+        }
+
+        let vicinity = evm_backend::Vicinity {
+            gas_price: gas_price_in,
+            origin: source,
+        };
+
+        let mut backend = evm_backend::Backend::<'_, C>::new(vicinity, ctx);
+        let metadata = StackSubstateMetadata::new(u64::max_value(), &Self::EVM_CONFIG);
```

Comment:
Why do you set the gas limit to `u64::max_value()` here? Shouldn't it be set to `gas_limit`? Because currently you only do the check after executing the transaction which opens up a DoS vector where people can execute stuff for which they have not paid.

Reply:
Fixed, thanks :)